### PR TITLE
Add validation check for username

### PR DESCRIPTION
**Pull request by SecondThundeR**
## Description

Fixes an error during the creation of a new user when the username is invalid

## Bugs and Issues

This issue was described in #168 

## How Has This Been Tested?

**Test Configuration**:
* Hardware: VirtualBox 6.1
* Specific steps: Ran installer and entered the username with incorrect case, characters and length

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code to the best of my abilities
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation where possible/if applicable

### CODE REVIEW

**Comment by pyfisch** [comment]
On a related note the complete regex from `useradd` should be checked: https://man.archlinux.org/man/useradd.8#CAVEATS

Personally I prefer to receive an error message if I enter a name containing upper-case letters over lowercasing the name.

**Comment by Torxed** [comment]
> On a related note the complete regex from `useradd` should be checked: https://man.archlinux.org/man/useradd.8#CAVEATS
> 
> Personally I prefer to receive an error message if I enter a name containing upper-case letters over lowercasing the name.

That's handy, so `[a-z_][a-z0-9_-]*[$]?` needs to be matched, otherwise we log an error.
Originally that was part of the discussion, so perhaps that's the route we should go.. It was me who said "just do `.lower()`" which is my bad. The regex would be nicer! And log an error and continue the loop.

**Comment by SecondThundeR** [comment]
Should we implement such a solution where we check if the username is invalid or its length is incorrect?
```python
if not re.match('[a-z_][a-z0-9_-]*[$]?', new_user) or len(new_user) > 32:
  log(
    "Invalid username. ..." # Add text to clarify the cause of the error
    level=LOG_LEVELS.Warning,
    fg='red'
  )
else:
  # Continuation of loop...
```

**Comment by Torxed** [comment]
> Should we implement such a solution where we check if the username is invalid or its length is incorrect?
> 
> ```python
> if not re.match('[a-z_][a-z0-9_-]*[$]?', new_user) or len(new_user) > 32:
>   log(
>     "Invalid username. ..." # Add text to clarify the cause of the error
>     level=LOG_LEVELS.Warning,
>     fg='red'
>   )
> else:
>   # Continuation of loop...
> ```

That looks very ideal!

**Comment by SecondThundeR** [comment]
Updated the code. Waiting for additional verification from your side

File: archinstall/lib/user_interaction.py [hunk 1]
```diff
@@ -49,8 +49,15 @@ def print_large_list(options, padding=5, margin_bottom=0, separator=': '):
 
 def ask_for_superuser_account(prompt='Create a required super-user with sudo privileges: ', forced=False):
 	while 1:
-		new_user = input(prompt).strip(' ').lower()
-		
+		new_user = input(prompt).strip(' ')
+
+		if not re.match('[a-z_][a-z0-9_-]*[$]?', new_user) or len(new_user) > 32:
```
Comment (kpcyrd) [inline archinstall/lib/user_interaction.py:54]: I think there's a `^` needed at the start.

File: archinstall/lib/user_interaction.py [hunk 2]
```diff
@@ -67,9 +74,16 @@ def ask_for_additional_users(prompt='Any additional users to install (leave blan
 	super_users = {}
 
 	while 1:
-		new_user = input(prompt).strip(' ').lower()
+		new_user = input(prompt).strip(' ')
 		if not new_user:
 			break
+		if not re.match('[a-z_][a-z0-9_-]*[$]?', new_user) or len(new_user) > 32:
```
Comment (kpcyrd) [inline archinstall/lib/user_interaction.py:80]: This code shouldn't be duplicated and live in a function instead.

File: archinstall/lib/user_interaction.py (hunk 1 above)
Comment (SecondThundeR) [reply archinstall/lib/user_interaction.py]: You're right. Updated this and fixed an issue where it accepts a username with a last character that is not a dollar sign

**Comment by Torxed** [comment]
This looks good to me, I'm ready to merge and will do so if no other objections :)

**Comment by kpcyrd** [comment]
btw, what does `[$]?` do? It seems this means "the end of the string or nothing" which means the regex doesn't actually force that it matches the whole string:
```
>>> re.match('^[a-z_][a-z0-9_-]*[$]?', 'asdf/_\n\x00')
<re.Match object; span=(0, 4), match='asdf'>
```
This would be fine if we're using the match, but we only check if the regex matched _somewhere_ and then use the whole string.

**Comment by Torxed** [comment]
> k if the regex matched _somewhere_ and then use the whole string.

I'm not a regexp ninja, but yea that looks wrong.
A quick search also shows that this is a Arch specific addon to upstream man page for useradd? In which case, perhaps https://archlinux.org/packages/core/x86_64/man-db/ should get a bug report as well.

**Comment by SecondThundeR** [comment]
> This would be fine if we're using the match, but we only check if the regex matched _somewhere_ and then use the whole string.

I tried to think about such a case in the future and added `$` at the end, which means meta-escape in the regular expression. Therefore, it's strictly checked whether there was a dollar sign at the very end or not. If the last character is different from it and isn't part of the previous rule `[a-z0-9_-]`, it returns `None`, but not the part of the string before that character. For other scenarios without this symbol, it is handled correctly. Since the dollar sign is optional, this should work fine.

```
>>> re.match('^[a-z_][a-z0-9_-]*\$?$', 'asdf/_\n\x00')
None

>>> re.match('^[a-z_][a-z0-9_-]*\$?$', 'asdf$')
<re.Match object; span=(0, 5), match='asdf$'>

>>> re.match('^[a-z_][a-z0-9_-]*\$?$', 'asdf123ds@')
None

>>> re.match('^[a-z_][a-z0-9_-]*\$?$', 'asdf123ds')
<re.Match object; span=(0, 9), match='asdf123ds'>
...
```
Also note that here I have used `\$` instead of `[$]`, which is equivalent. Thought it would be better not to use square brackets for single character

**Pull request merged**